Long form variable declaration and assignment

The following program fails:

```go
package main

import (
	"external"
)

func main() {
	var i int = 10
	external.Printf("%d\n", i)
}

```

Trace:

```
2019/12/26 21:06:33 unable to handle default: 21 - {Type:OPERATOR, Val:=, Line:8}
panic: unable to handle default: 21 - {Type:OPERATOR, Val:=, Line:8}

goroutine 1 [running]:
log.Panicf(0x11c79dd, 0x22, 0xc00011f588, 0x2, 0x2)
	/Users/gustav/src/go/src/log/log.go:345 +0xc0
github.com/zegl/tre/compiler/parser.(*parser).parseOneWithOptions(0xc00011fb28, 0x10101, 0xc0000b2060, 0xc0000f8040)
	/Users/gustav/src/tre/compiler/parser/parser.go:589 +0x2508
github.com/zegl/tre/compiler/parser.(*parser).parseOne(0xc00010fb28, 0x1, 0x0, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:49 +0x3c
github.com/zegl/tre/compiler/parser.(*parser).parseUntilEither(0xc00011fb28, 0xc00011f728, 0x1, 0x1, 0x0, 0x8, 0x20300000000000, 0xc0000b2030, 0xc0000f4070, 0xc000052748, ...)
	/Users/gustav/src/tre/compiler/parser/parser.go:966 +0x16f
github.com/zegl/tre/compiler/parser.(*parser).parseUntil(0xc000052b28, 0x5, 0x11c0a3a, 0x1, 0x0, 0x20300000000000, 0x16fffff, 0xc000052810)
	/Users/gustav/src/tre/compiler/parser/parser.go:942 +0x97
github.com/zegl/tre/compiler/parser.(*parser).parseOneWithOptions(0xc00011fb28, 0x10101, 0x0, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:447 +0xec3
github.com/zegl/tre/compiler/parser.(*parser).parseOne(0xc000052b28, 0xc0000b2001, 0x0, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:49 +0x3c
github.com/zegl/tre/compiler/parser.(*parser).parseUntilEither(0xc00011fb28, 0xc00011fab0, 0x1, 0x1, 0xc000052af0, 0x8402b31d, 0xbbcd90b3491ffdd5, 0x100d648, 0xbb000000011855c0, 0xc000106001, ...)
	/Users/gustav/src/tre/compiler/parser/parser.go:966 +0x16f
github.com/zegl/tre/compiler/parser.(*parser).parseUntil(0xc000052b28, 0xc000106006, 0x0, 0x0, 0x0, 0x18, 0xc0000f41fa, 0xc0000f4280)
	/Users/gustav/src/tre/compiler/parser/parser.go:942 +0x97
github.com/zegl/tre/compiler/parser.Parse(0xc000104000, 0x59, 0x80, 0x1, 0x80, 0x109, 0x0)
	/Users/gustav/src/tre/compiler/parser/parser.go:44 +0x2e6
github.com/zegl/tre/cmd/tre/build.parseFile(0x7ffeefbffa99, 0x1a, 0x11ff9a0, 0xc0000b4f70, 0x0)
	/Users/gustav/src/tre/cmd/tre/build/build.go:177 +0xbc
github.com/zegl/tre/cmd/tre/build.compilePackage(0xc0000e2140, 0x7ffeefbffa99, 0x1a, 0xc0000c2020, 0x19, 0x11c0d00, 0x4, 0x19, 0xc000052e80)
	/Users/gustav/src/tre/cmd/tre/build/build.go:103 +0xbc
github.com/zegl/tre/cmd/tre/build.Build(0x7ffeefbffa99, 0x1a, 0xc0000c2020, 0x19, 0x7ffeefbffaab, 0x5, 0x1, 0xc000052f10, 0x19)
	/Users/gustav/src/tre/cmd/tre/build/build.go:24 +0x9f
main.main()
	/Users/gustav/src/tre/cmd/tre/main.go:48 +0x2fd
```

--- a/compiler/compiler/alloc.go
+++ b/compiler/compiler/alloc.go
@@ -21,51 +21,49 @@ func (c *Compiler) compileAllocNode(v *parser.AllocNode) {
 	}()
 
 	// Allocate from type
-	if len(v.Val) == 1 {
-		if typeNode, ok := v.Val[0].(parser.TypeNode); ok {
-			treType := c.parserTypeToType(typeNode)
+	if len(v.Val) == 0 && v.Type != nil {
+		treType := c.parserTypeToType(v.Type)
 
-			var val llvmValue.Value
-			var block *ir.Block
+		var val llvmValue.Value
+		var block *ir.Block
 
-			// Package level variables
-			if c.contextBlock == nil {
-				globType := treType.LLVM()
-				glob := c.module.NewGlobal(name.Var(v.Name[0]), globType)
-				glob.Init = constant.NewZeroInitializer(globType)
+		// Package level variables
+		if c.contextBlock == nil {
+			globType := treType.LLVM()
+			glob := c.module.NewGlobal(name.Var(v.Name[0]), globType)
+			glob.Init = constant.NewZeroInitializer(globType)
 
-				c.currentPackage.DefinePkgVar(v.Name[0], value.Value{
-					Value:      glob,
-					Type:       treType,
-					IsVariable: true,
-				})
+			c.currentPackage.DefinePkgVar(v.Name[0], value.Value{
+				Value:      glob,
+				Type:       treType,
+				IsVariable: true,
+			})
 
-				val = glob
-				block = c.initGlobalsFunc.Blocks[0]
-			} else {
-				alloc := c.contextBlock.NewAlloca(treType.LLVM())
-				alloc.SetName(name.Var(v.Name[0]))
+			val = glob
+			block = c.initGlobalsFunc.Blocks[0]
+		} else {
+			alloc := c.contextBlock.NewAlloca(treType.LLVM())
+			alloc.SetName(name.Var(v.Name[0]))
 
-				c.setVar(v.Name[0], value.Value{
-					Value:      alloc,
-					Type:       treType,
-					IsVariable: true,
-				})
+			c.setVar(v.Name[0], value.Value{
+				Value:      alloc,
+				Type:       treType,
+				IsVariable: true,
+			})
 
-				val = alloc
-				block = c.contextBlock
-			}
-
-			// Set to zero values
-			// TODO: Make slices less special
-			if sliceType, ok := treType.(*types.Slice); ok {
-				sliceType.SliceZero(block, c.externalFuncs.Malloc.Value.(llvmValue.Named), 2, val)
-			} else {
-				treType.Zero(block, val)
-			}
-
-			return
+			val = alloc
+			block = c.contextBlock
 		}
+
+		// Set to zero values
+		// TODO: Make slices less special
+		if sliceType, ok := treType.(*types.Slice); ok {
+			sliceType.SliceZero(block, c.externalFuncs.Malloc.Value.(llvmValue.Named), 2, val)
+		} else {
+			treType.Zero(block, val)
+		}
+
+		return
 	}
 
 	for valIndex, valNode := range v.Val {

--- a/compiler/compiler/compiler.go
+++ b/compiler/compiler/compiler.go
@@ -219,6 +219,10 @@ func (c *Compiler) compile(instructions []parser.Node) {
 			c.compileReturnNode(v)
 		case *parser.AllocNode:
 			c.compileAllocNode(v)
+		case *parser.AllocGroup:
+			for _, a := range v.Allocs {
+				c.compileAllocNode(a)
+			}
 		case *parser.AssignNode:
 			c.compileAssignNode(v)
 		case *parser.ForNode:

--- a/compiler/lexer/lexer.go
+++ b/compiler/lexer/lexer.go
@@ -218,7 +218,7 @@ func Lex(inputFullSource string) []Item {
 		}
 	}
 
-	res = append(res, Item{Type: EOF})
+	res = append(res, Item{Type: EOL}, Item{Type: EOF})
 
 	return res
 }

--- a/compiler/lexer/lexer_test.go
+++ b/compiler/lexer/lexer_test.go
@@ -13,7 +13,8 @@ func TestLexerSimpleAdd(t *testing.T) {
 		{Type: IDENTIFIER, Val: "aa", Line: 1},
 		{Type: OPERATOR, Val: "+", Line: 1},
 		{Type: IDENTIFIER, Val: "b", Line: 1},
-		{Type: EOF, Val: ""},
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -33,7 +34,8 @@ func TestLexerSimpleAddWithNewlines(t *testing.T) {
 		{Type: OPERATOR, Val: "+", Line: 2},
 		{Type: IDENTIFIER, Val: "b", Line: 2},
 
-		{Type: EOF, Val: ""},
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -46,7 +48,9 @@ func TestLexerSimpleAddNumber(t *testing.T) {
 		{Type: IDENTIFIER, Val: "aa", Line: 1},
 		{Type: OPERATOR, Val: "+", Line: 1},
 		{Type: NUMBER, Val: "14", Line: 1},
-		{Type: EOF, Val: ""},
+
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -60,7 +64,9 @@ func TestLexerSimpleCall(t *testing.T) {
 		{Type: OPERATOR, Val: "(", Line: 1},
 		{Type: IDENTIFIER, Val: "bar", Line: 1},
 		{Type: OPERATOR, Val: ")", Line: 1},
-		{Type: EOF, Val: ""},
+
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -74,7 +80,9 @@ func TestLexerSimpleCallWithString(t *testing.T) {
 		{Type: OPERATOR, Val: "(", Line: 1},
 		{Type: STRING, Val: "bar", Line: 1},
 		{Type: OPERATOR, Val: ")", Line: 1},
-		{Type: EOF, Val: ""},
+
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -85,7 +93,8 @@ func TestString(t *testing.T) {
 
 	expected := []Item{
 		{Type: STRING, Val: "bar", Line: 1},
-		{Type: EOF, Val: ""},
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -96,7 +105,8 @@ func TestEscapedString(t *testing.T) {
 
 	expected := []Item{
 		{Type: STRING, Val: "bar\"", Line: 1},
-		{Type: EOF, Val: ""},
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -112,7 +122,8 @@ func TestLexerSimpleCallWithTwoStrings(t *testing.T) {
 		{Type: OPERATOR, Val: ",", Line: 1},
 		{Type: STRING, Val: "baz", Line: 1},
 		{Type: OPERATOR, Val: ")", Line: 1},
-		{Type: EOF, Val: ""},
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)
@@ -128,7 +139,8 @@ func TestLexerSimpleCallWithStringNum(t *testing.T) {
 		{Type: OPERATOR, Val: ",", Line: 1},
 		{Type: NUMBER, Val: "123", Line: 1},
 		{Type: OPERATOR, Val: ")", Line: 1},
-		{Type: EOF, Val: ""},
+		{Type: EOL},
+		{Type: EOF},
 	}
 
 	assert.Equal(t, expected, r)

--- a/compiler/parser/node.go
+++ b/compiler/parser/node.go
@@ -235,10 +235,21 @@ type AllocNode struct {
 
 	Name []string
 	Val  []Node
+
+	Type TypeNode // Is set when allocating on the format "var ident int" and "var ident, ident int = expr, expr"
 }
 
 func (an AllocNode) String() string {
-	return fmt.Sprintf("alloc(%s) = %v (escapes: %v)", an.Name, an.Val, an.Escapes)
+	return fmt.Sprintf("alloc(%s) = %v (escapes: %v type: %v)", an.Name, an.Val, an.Escapes, an.Type)
+}
+
+type AllocGroup struct {
+	baseNode
+	Allocs []*AllocNode
+}
+
+func (an AllocGroup) String() string {
+	return fmt.Sprintf("alloc(%+v)", an.Allocs)
 }
 
 // AssignNode assign Val to Target (or Name)

--- a/compiler/parser/parser.go
+++ b/compiler/parser/parser.go
@@ -530,22 +530,26 @@ func (p *parser) parseOneWithOptions(withAheadParse, withArithAhead, withIdentif
 		if current.Val == "var" {
 			p.i++
 
-			name := p.lookAhead(0)
-			if name.Type != lexer.IDENTIFIER {
-				panic("expected IDENTIFIER after var")
+			isGroup := p.lookAhead(0)
+			if isGroup.Val == "(" {
+				p.i++
+
+				var allocs []*AllocNode
+				for {
+					nextToken := p.lookAhead(0)
+					if nextToken.Val == ")" {
+						break
+					}
+					if nextToken.Type == lexer.EOL {
+						p.i++
+						continue
+					}
+					allocs = append(allocs, p.parseVarDecl())
+				}
+				return &AllocGroup{Allocs: allocs}
 			}
 
-			p.i++
-
-			tp, err := p.parseOneType()
-			if err != nil {
-				panic(err)
-			}
-
-			return &AllocNode{
-				Name: []string{name.Val},
-				Val:  []Node{tp},
-			}
+			return p.parseVarDecl()
 		}
 
 		if current.Val == "package" {
@@ -606,6 +610,65 @@ func (p *parser) parseOneWithOptions(withAheadParse, withArithAhead, withIdentif
 	p.printInput()
 	log.Panicf("unable to handle default: %d - %+v", p.i, current)
 	panic("")
+}
+
+func (p *parser) parseVarDecl() *AllocNode {
+	allocNode := &AllocNode{Name: p.identifierList()}
+
+	isEq := p.lookAhead(0)
+	if isEq.Type == lexer.OPERATOR && isEq.Val == "=" {
+		p.i++
+		allocNode.Val = p.expressionList()
+		return allocNode
+	} else {
+		tp, err := p.parseOneType()
+		if err != nil {
+			panic(err)
+		}
+		allocNode.Type = tp
+		p.i++
+	}
+
+	isEq = p.lookAhead(0)
+	if isEq.Type == lexer.OPERATOR && isEq.Val == "=" {
+		p.i++
+		allocNode.Val = p.expressionList()
+	}
+
+	return allocNode
+}
+
+func (p *parser) identifierList() (res []string) {
+	for {
+		n := p.lookAhead(0)
+		p.expect(n, lexer.Item{Type: lexer.IDENTIFIER})
+		res = append(res, n.Val)
+
+		p.i++
+
+		isComma := p.lookAhead(0)
+		if isComma.Type == lexer.OPERATOR && isComma.Val == "," {
+			p.i++
+			continue
+		}
+
+		return
+	}
+}
+
+func (p *parser) expressionList() (res []Node) {
+	for {
+		res = append(res, p.parseOne(true))
+		p.i++
+
+		isComma := p.lookAhead(0)
+		if isComma.Type == lexer.OPERATOR && isComma.Val == "," {
+			p.i++
+			continue
+		}
+
+		return
+	}
 }
 
 func (p *parser) aheadParse(input Node) Node {

--- a/compiler/parser/var_alloc_test.go
+++ b/compiler/parser/var_alloc_test.go
@@ -1,0 +1,100 @@
+package parser
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+
+	"github.com/zegl/tre/compiler/lexer"
+)
+
+func TestAllocType(t *testing.T) {
+	lexed := lexer.Lex(`var a int`)
+
+	expected := FileNode{
+		Instructions: []Node{
+			&AllocNode{
+				Name: []string{"a"},
+				Type: &SingleTypeNode{TypeName: "int"},
+			},
+		},
+	}
+
+	assert.Equal(t, expected, Parse(lexed, false))
+}
+
+func TestAllocTypeWithValue(t *testing.T) {
+	lexed := lexer.Lex(`var a int = 10`)
+
+	expected := FileNode{
+		Instructions: []Node{
+			&AllocNode{
+				Name: []string{"a"},
+				Type: &SingleTypeNode{TypeName: "int"},
+				Val:  []Node{&ConstantNode{Type: NUMBER, Value: 10}},
+			},
+		},
+	}
+
+	assert.Equal(t, expected, Parse(lexed, false))
+}
+
+func TestAllocImplicitTypeValue(t *testing.T) {
+	lexed := lexer.Lex(`var a = 10`)
+
+	expected := FileNode{
+		Instructions: []Node{
+			&AllocNode{
+				Name: []string{"a"},
+				Val:  []Node{&ConstantNode{Type: NUMBER, Value: 10}},
+			},
+		},
+	}
+
+	assert.Equal(t, expected, Parse(lexed, false))
+}
+
+func TestAllocMultiWithType(t *testing.T) {
+	lexed := lexer.Lex(`var a, b int`)
+
+	expected := FileNode{
+		Instructions: []Node{
+			&AllocNode{
+				Name: []string{"a", "b"},
+				Type: &SingleTypeNode{TypeName: "int"},
+			},
+		},
+	}
+
+	assert.Equal(t, expected, Parse(lexed, false))
+}
+
+func TestAllocGroup(t *testing.T) {
+	lexed := lexer.Lex(`var (
+	a int
+	b, c uint8
+	d = 10
+)`)
+
+	expected := FileNode{
+		Instructions: []Node{
+			&AllocGroup{
+				Allocs: []*AllocNode{
+					{
+						Name: []string{"a"},
+						Type: &SingleTypeNode{TypeName: "int"},
+					},
+					{
+						Name: []string{"b", "c"},
+						Type: &SingleTypeNode{TypeName: "uint8"},
+					},
+					{
+						Name: []string{"d"},
+						Val:  []Node{&ConstantNode{Type: NUMBER, Value: 10}},
+					},
+				},
+			},
+		},
+	}
+
+	assert.Equal(t, expected, Parse(lexed, false))
+}

--- a/compiler/testdata/variable.go
+++ b/compiler/testdata/variable.go
@@ -3,9 +3,29 @@ package main
 import "external"
 
 func main() {
+	var (
+		d    int
+		e, f int = 30, 40
+		g, h     = 50, 60
+	)
+
+	external.Printf("d is = %d\n", d) // d is = 0
+	external.Printf("e is = %d\n", e) // e is = 30
+	external.Printf("f is = %d\n", f) // f is = 40
+	external.Printf("g is = %d\n", g) // g is = 50
+	external.Printf("h is = %d\n", h) // h is = 60
+
 	foo := 4
 	foo = foo + 5
 	foo = foo + 6 + 7 + 8
-	// foo is = 30
-	external.Printf("foo is = %d\n", foo)
+	external.Printf("foo is = %d\n", foo) // foo is = 30
+
+	var a int
+	external.Printf("a is = %d\n", a) // a is = 0
+
+	var b int = 20
+	external.Printf("b is = %d\n", b) // b is = 20
+
+	var c = 21
+	external.Printf("c is = %d\n", c) // c is = 21
 }

